samples/Cordova/Backend/Providers/AuthorizationProvider.cs

```diff
@@ -72,6 +74,68 @@ public sealed class AuthorizationProvider : OpenIdConnectServerProvider {
+                context.Reject(
+                    error: OpenIdConnectConstants.Errors.UnsupportedGrantType,
+                    description: "Only authorization code and refresh token grant types " +
+                                 "are accepted by this authorization server");
+
+                return;
+            }
+
+            // Note: client authentication is not mandatory for non-confidential client applications like mobile apps
+            // (except when using the client credentials grant type) but this authorization server uses a safer policy
+            // that makes client authentication mandatory and returns an error if client_id or client_secret is missing.
+            // You may consider relaxing it to support the resource owner password credentials grant type
+            // with JavaScript or desktop applications, where client credentials cannot be safely stored.
+            // In this case, call context.Skip() to inform the server middleware the client is not trusted.
+            if (string.IsNullOrEmpty(context.ClientId) || string.IsNullOrEmpty(context.ClientSecret))
+            {
+                //context.Reject(
+                //    error: OpenIdConnectConstants.Errors.InvalidRequest,
+                //    description: "Missing credentials: ensure that your credentials were correctly " +
+                //                 "flowed in the request body or in the authorization header");
+                context.Skip();
+                return;
+            }
+
+            var database = context.HttpContext.RequestServices.GetRequiredService<ApplicationContext>();
+
+            // Retrieve the application details corresponding to the requested client_id.
+            var application = await (from entity in database.Applications
+                                     where entity.ApplicationID == context.ClientId
+                                     select entity).SingleOrDefaultAsync(context.HttpContext.RequestAborted);
+
+            if (application == null) {
+                context.Reject(
+                    error: OpenIdConnectConstants.Errors.InvalidClient,
+                    description: "Application not found in the database: ensure that your client_id is correct");
+
+                return;
+            }
+
+            // Note: to mitigate brute force attacks, you SHOULD strongly consider applying
+            // a key derivation function like PBKDF2 to slow down the secret validation process.
+            // You SHOULD also consider using a time-constant comparer to prevent timing attacks.
+            // For that, you can use the CryptoHelper library developed by @henkmollema:
+            // https://github.com/henkmollema/CryptoHelper. If you don't need .NET Core support,
+            // SecurityDriven.NET/inferno is a rock-solid alternative: http://securitydriven.net/inferno/
+            if (!string.Equals(context.ClientSecret, application.Secret, StringComparison.Ordinal))
```

Comment:
I can just remove almost everything (except first flow test) and just put context.Skip() then?